libraries/AP_Frsky_Telem/mavlite.cpp

```diff
@@ -0,0 +1,288 @@
+            break;
+        case PARSE_STATE_GOT_START:
+            if ( offset == 0 && byte == 0x00 ) {
+                mavlite_init_parse(rxmsg, status);
+                status->parse_state = PARSE_STATE_GOT_START;
+            } else {
+                rxmsg->len = byte;
+                status->parse_state = PARSE_STATE_GOT_LEN;
+            }
+            mavlite_update_checksum(rxmsg, byte);
+            break;
+        case PARSE_STATE_GOT_LEN:
+            if ( offset == 0 && byte == 0x00 ) {
+                mavlite_init_parse(rxmsg, status);
+                status->parse_state = PARSE_STATE_GOT_START;
+            } else {
+                rxmsg->msgid = byte;
+                status->parse_state = PARSE_STATE_GOT_MSGID;
+            }
+            mavlite_update_checksum(rxmsg, byte);
+            break;
+        case PARSE_STATE_GOT_MSGID:
+            if ( offset == 0 && byte == 0x00 ) {
+                mavlite_init_parse(rxmsg, status);
+                status->parse_state = PARSE_STATE_GOT_START;
+            } else {
+                rxmsg->payload[status->payload_next_byte++] = byte;
+                status->parse_state = PARSE_STATE_GOT_PAYLOAD;
+            }
+            mavlite_update_checksum(rxmsg, byte);
+            break;
+        case PARSE_STATE_GOT_SEQ:
+            if ( offset == 0 && byte == 0x00 ) {
+                mavlite_init_parse(rxmsg, status);
+                status->parse_state = PARSE_STATE_GOT_START;
+                mavlite_update_checksum(rxmsg, byte);
+            } else {
+                if ( status->payload_next_byte < rxmsg->len ) {
+                    rxmsg->payload[status->payload_next_byte++] = byte;
+                    status->parse_state = PARSE_STATE_GOT_PAYLOAD;
+                    mavlite_update_checksum(rxmsg, byte);
+                } else {
+                    if ( rxmsg->checksum == byte ) {
+                        status->parse_state = PARSE_STATE_MESSAGE_RECEIVED;
+                        return true;
+                    } else {
+                        status->parse_state = PARSE_STATE_ERROR;
+                    }
+                }
+            }
+            break;
+        case PARSE_STATE_GOT_PAYLOAD:
+            if ( offset == 0) {
+                if ( byte == 0x00 ) {
+                    mavlite_init_parse(rxmsg, status);
+                    status->parse_state = PARSE_STATE_GOT_START;
+                } else {
+                    if ((byte & 0x3F) != status->current_rx_seq + 1) {
+                        status->parse_state = PARSE_STATE_ERROR;
+                    } else {
+                        status->current_rx_seq = (byte & 0x3F);
+                        status->parse_state = PARSE_STATE_GOT_SEQ;
+                    }
+                }
+                mavlite_update_checksum(rxmsg, byte);
+            } else {
+                if ( status->payload_next_byte < rxmsg->len ) {
+                    rxmsg->payload[status->payload_next_byte++] = byte;
```

Comment:
We need bounds-checking here.  If `rxmsg->len` is 255 then we've got a substantial memory-overwrite problem here.